c10/core/Event.h

```diff
@@ -0,0 +1,145 @@
+    moveHelper(std::move(other));
+    return *this;
+}
+
+void moveHelper(Event&& other) {
+    std::swap(was_recorded_, other.was_recorded_);
+    std::swap(flags_, other.flags_);
+    std::swap(stream_, other.stream_);
+    std::swap(event_, other.event_);
+}
+
+// Destructor
+~Event() {
+    if (stream_) {
+        switch(stream_->device_type()) {
+            case DeviceType::CUDA:
+                at::detail::getCUDAHooks().destroyCUDAEvent(this);
+            default:
+                break; // no throw
+        }
+    }
+}
+
+// Enqueues this event in the given stream to be marked as completed when
+// the stream processes the event.
+void record(const Stream& stream) {
+    if (stream_) {
+        TORCH_CHECK(
+            stream_->device_type() == stream.device_type()
+        ,   "Event device type", DeviceTypeName(stream_->device_type())
+        ,   "does not match recording stream's device type", DeviceTypeName(stream.device_type())
+        ,   ".");
+    }
+
+    switch(stream.device_type()) {
+        case DeviceType::CUDA:
+            at::detail::getCUDAHooks().recordCUDAEvent(this, stream);
+            break;
+        default:
+            TORCH_CHECK(false, "Device type", DeviceTypeName(stream.device_type()), "does not support recording events.");
+    }
+
+    was_recorded_ = true;
+    stream_ = stream;
+}
+
+// Records the event if (and only if) it has not been previously recorded.
+void recordOnce(const Stream& stream) {
+    if (!was_recorded_) record(stream);
+}
+
+// Causes the given stream to wait for the last event recorded to be
+// completed (by the stream it was recorded on). 
+void block(const Stream& stream) const {
+    if (stream_) {
+        TORCH_CHECK(
+            stream_->device_type() == stream.device_type()
+        ,   "Event device type", DeviceTypeName(stream_->device_type())
+        ,   "does not match stream to block's device type", DeviceTypeName(stream.device_type())
+        ,   ".");
+
+        switch(stream.device_type()) {
+            case DeviceType::CUDA:
+                at::detail::getCUDAHooks().blockCUDAStream(*this, stream);
+                break;
+            default:
+                TORCH_CHECK(false, "Device type", DeviceTypeName(stream.device_type()), "does not support blocking on events.");
+        }
+    }
+}
+
+// Returns false if the last recorded event exists has not yet completed,
+// and true otherwise.
+bool query() const {
+    if (!stream_) return true;
+
+    switch(stream_->device_type()) {
+        case DeviceType::CUDA:
+            return at::detail::getCUDAHooks().queryCUDAEvent(*this);
+        default:
+            TORCH_CHECK(false, "Device type", DeviceTypeName(stream_->device_type()), "does not support querying events.");
+    }
+
+    return false;
+}
+
+unsigned int flags() const noexcept { return flags_; }
+
+private: 
+    bool was_recorded_ = false;
+    unsigned int flags_;
+    optional<Stream> stream_ = nullopt;
+    void* event_ = nullptr;
```

Comment:
Is it possible to make this a `unique_ptr` with a function pointer deleter instead? It seems marginally more safe.

Reply:
Good catch. Yes, we can use a unique_ptr with a custom deleter here.